samples/boards/nrf/s2ram/src/main.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2022 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <pm/pm.h>
+#include <pm/policy.h>
+
+#include <arch/common/pm_s2ram.h>
+
+#include <hal/nrf_vmc.h>
+#include <hal/nrf_power.h>
+#include <hal/nrf_regulators.h>
+#include <hal/nrf_gpio.h>
+
+#define RAM_N_BLOCK	(8)
+#define MASK_ALL_SECT	(VMC_RAM_POWER_S0RETENTION_Msk  | VMC_RAM_POWER_S1RETENTION_Msk  | \
+			 VMC_RAM_POWER_S2RETENTION_Msk  | VMC_RAM_POWER_S3RETENTION_Msk  | \
+			 VMC_RAM_POWER_S4RETENTION_Msk  | VMC_RAM_POWER_S5RETENTION_Msk  | \
+			 VMC_RAM_POWER_S6RETENTION_Msk  | VMC_RAM_POWER_S7RETENTION_Msk  | \
+			 VMC_RAM_POWER_S8RETENTION_Msk  | VMC_RAM_POWER_S9RETENTION_Msk  | \
+			 VMC_RAM_POWER_S10RETENTION_Msk | VMC_RAM_POWER_S11RETENTION_Msk | \
+			 VMC_RAM_POWER_S12RETENTION_Msk | VMC_RAM_POWER_S13RETENTION_Msk | \
+			 VMC_RAM_POWER_S14RETENTION_Msk | VMC_RAM_POWER_S15RETENTION_Msk)
+
+static void ram_retention_config(bool enable)
+{
+	uint32_t mask_sect = MASK_ALL_SECT;
+
+	for (size_t n = 0; n < RAM_N_BLOCK; n++) {
+		if (enable) {
+			nrf_vmc_ram_block_retention_set(NRF_VMC, n, mask_sect);
+		} else {
+			nrf_vmc_ram_block_retention_clear(NRF_VMC, n, mask_sect);
+		}
+	}
+}
+
+static void system_off(void)
+{
+	nrf_regulators_system_off(NRF_REGULATORS);
+}
+
+static void do_suspend(void)
+{
+	irq_lock();
+
+	/*
+	 * Save the CPU context (including the return address),set the SRAM
+	 * marker and power off the system.
+	 */
+	arch_pm_s2ram_suspend(system_off);
+
+	/*
+	 * XXX: On resuming we return exactly *HERE*
+	 */
+}
+
+void main(void)
+{
+	printk("S2RAM demo (%s)\n", CONFIG_BOARD);
+
+	/*
+	 * Enable RAM retention for *ALL* the SRAM
+	 */
+	ram_retention_config(true);
+
+	/*
+	 * Enable wake-up button and LED1
+	 */
+	nrf_gpio_cfg_input(23,  NRF_GPIO_PIN_PULLUP);
+	nrf_gpio_cfg_sense_set(23, NRF_GPIO_PIN_SENSE_LOW);
+	nrf_gpio_cfg_output(28);
+
+	/*
+	 * Sleep enough time to trigger S2RAM
+	 */
+	while (true) {
+		k_msleep(1000);
+	}
+}
+
+void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	if (state != PM_STATE_SUSPEND_TO_RAM) {
+		return;
+	}
+
+	do_suspend();
+}
+
+void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
```

Comment:
this is tipically SoC code (`soc/`), any reason it sits on the application here?

Reply:
Because this is a sample showing off how this should be implemented. All the pieces around resuming the other core peripherals (NVIC, MPU, etc...) are missing so it is definitely not ready to be upgraded to a code worthy to be in `soc/` (yet).